Index: src/main.c
===================================================================
RCS file: /cvsroot/pcb/pcb/src/main.c,v
retrieving revision 1.60
diff -U3 -p -r1.60 main.c
--- src/main.c	21 Oct 2006 14:38:30 -0000	1.60
+++ src/main.c	1 Dec 2006 00:04:52 -0000
@@ -55,6 +55,10 @@
 /* This next one is so we can print the help messages. */
 #include "hid/hidint.h"
 
+#ifdef HAVE_DBUS
+#include "dbus.h"
+#endif
+
 #ifdef HAVE_LIBDMALLOC
 #include <dmalloc.h>
 #endif
@@ -767,7 +771,13 @@ main (int argc, char *argv[])
   if (Settings.init_done == 0)
     {
       Settings.init_done = 1;
+#if HAVE_DBUS
+      pcb_dbus_setup();
+#endif
       gui->do_export (0);
+#if HAVE_DBUS
+      pcb_dbus_finish();
+#endif
     }
 
   return (0);
